chpl/chpl-service/src/main/java/gov/healthit/chpl/upload/listing/handler/TestToolUploadHandler.java

```diff
@@ -0,0 +1,77 @@
+package gov.healthit.chpl.upload.listing.handler;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.commons.csv.CSVRecord;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import gov.healthit.chpl.domain.CertificationResultTestTool;
+import gov.healthit.chpl.upload.listing.Headings;
+import gov.healthit.chpl.upload.listing.ListingUploadHandlerUtil;
+
+@Component("testToolUploadHandler")
+public class TestToolUploadHandler {
+    private ListingUploadHandlerUtil uploadUtil;
+
+    @Autowired
+    public TestToolUploadHandler(ListingUploadHandlerUtil uploadUtil) {
+        this.uploadUtil = uploadUtil;
+    }
+
+    public List<CertificationResultTestTool> handle(CSVRecord certHeadingRecord, List<CSVRecord> certResultRecords) {
+        List<CertificationResultTestTool> testTools = new ArrayList<CertificationResultTestTool>();
+        List<String> testToolNames = parseTestToolNames(certHeadingRecord, certResultRecords);
+        List<String> testToolVersions = parseTestToolVersions(certHeadingRecord, certResultRecords);
+        if (uploadUtil.areCollectionsEmpty(testToolNames, testToolVersions)) {
+            return testTools;
+        }
+
+        int max = 0;
+        if (CollectionUtils.isNotEmpty(testToolNames)) {
+            max = Math.max(max, testToolNames.size());
+        }
+        if (CollectionUtils.isNotEmpty(testToolVersions)) {
+            max = Math.max(max, testToolVersions.size());
+        }
+        //I think everything remains ordered using these data structures so this should be okay.
+        testTools = IntStream.range(0, max)
+            .mapToObj(index -> buildTestTool(index, testToolNames, testToolVersions))
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList());
```

Comment:
I'm not sure there is any guarantee to the processing order of a stream.  If you need them to be in a certain order, maybe use the `.sorted()` function.

Reply:
My understanding is that if the collection backing the stream is ordered (ArrayList, LinkedHashSet, etc) then the items in the stream will be encountered in the order they exist in the collection. But if the collection backing the stream is not ordered (HashSet) then there are no guarantees. That's just what I got from a few Google searches. If you get a few minutes to look into it please let me know if you agree or if you really think this code won't work. I use something like it all over the place and it *seems* to work correctly but I would not like to be wrong. :)